### DISCUSSION TITLE
Is there a way to enable HRTIM on F334?

### DISCUSSION THREAD

**Discussion by Stefankante** [stm32duino/Arduino_Core_STM32#1947] [Q&A] [answered]
Please give me some light.

I wanted to use HRTIM feature of this chip to generate high freq (>30000Hz) 16 bit PWM

**Comment by fpistm** [comment]
Hi,
you can try using HAL or LL.
Some examples are available here:
https://github.com/STMicroelectronics/STM32CubeF3/tree/master/Projects/STM32F3348-Discovery/Examples/HRTIM

and here you could find an example how convert a Cube example to Arduino sketch:
https://www.stm32duino.com/viewtopic.php?t=110

**Reply by Stefankante** [reply]
I wonder if there's easier solution like snippet. I can't wrap my mind around the one you send me.

**Reply by Stefankante** [reply]
Is there any hack to make use of HRTIM with analogWrite?

**Reply by fpistm** [reply]
Never try HRTIM, that's why I proposed you to use HAL at sketch level.

**Reply by Stefankante** [reply]
https://www.youtube.com/watch?v=b0M7hRunN7Y I saw this video but it was in flowcode.

**Comment by Stefankante** [comment, marked as answer]
I tried with some guide from here [https://sudonull.com/post/7101-Development-of-buck-converter-on-STM32F334-principle-of-operation-calculations-prototyping](url)

and strip out all ADC related only PWM part remain, Tested LED works on PA8 and can be control with 
`HRTIM1->sTimerxRegs[0].CMP1xR = dutycycle;` // 0-100% 

But I have no idea how to control in style of analogWrite which only need to stated Frequency,Resolution,Level in that resolution.
Also does Generic STM32F334R8xx board setting make use of HSE on board? 


#pragma once
#include "stm32f3xx.h"

void InitHRPWM (void);
void SetDutyTimerA (uint16_t duty);
void InitBasicADC (void);
extern void Delay (uint32_t setDelay);

uint16_t dutyControl = 0;
#define PeriodTimerA 	((uint16_t)45000)		// f = 102 kHz
/*********************************** Used functions *************************************/

void InitHRPWM (void) {

	RCC->CFGR3   |= RCC_CFGR3_HRTIM1SW_PLL;
	RCC->APB2ENR |= RCC_APB2ENR_HRTIM1EN;

	/************************************************
	 *                Setting GPIO
	 ***********************************************/

	RCC->AHBENR  |= RCC_AHBENR_GPIOAEN;
	GPIOA->MODER   &= ~GPIO_MODER_MODER8;
	GPIOA->MODER   |= GPIO_MODER_MODER8_1;																		
// Alternative PP
	GPIOA->OSPEEDR |= GPIO_OSPEEDER_OSPEEDR8;																	
// Very high speed
	GPIOA->MODER   &= ~GPIO_MODER_MODER9;
	GPIOA->MODER   |= GPIO_MODER_MODER9_1;
	GPIOA->OSPEEDR |= GPIO_OSPEEDER_OSPEEDR9;
	GPIOA->AFR[1] |= 0xDD;																						
// PA8 and PA9 - AF13

	/************************************************
	 *               Setting timer A
	 ***********************************************/

	HRTIM1->sCommonRegs.DLLCR |= HRTIM_DLLCR_CAL | HRTIM_DLLCR_CALEN;
	while ((HRTIM1->sCommonRegs.ISR & HRTIM_ISR_DLLRDY) == RESET);

	HRTIM1->sTimerxRegs[0].PERxR = PeriodTimerA;																
// Period for timer A
	HRTIM1->sTimerxRegs[0].CMP1xR = 0;																			   
 // Duty for timer A

	HRTIM1->sTimerxRegs[0].OUTxR |= HRTIM_OUTR_DTEN;															
// Deadtime enable
	HRTIM1->sTimerxRegs[0].DTxR  |= HRTIM_DTR_DTPRSC_0 | HRTIM_DTR_DTPRSC_1;									
// Tdtg = 6.94 ns
	HRTIM1->sTimerxRegs[0].DTxR  |= HRTIM_DTR_DTR_0 | HRTIM_DTR_DTR_1 | HRTIM_DTR_DTR_2 | HRTIM_DTR_DTR_3;		
// Deadtime rising = 15*Ttg = 104 ns
	HRTIM1->sTimerxRegs[0].DTxR  |= HRTIM_DTR_DTF_0 | HRTIM_DTR_DTF_1 | HRTIM_DTR_DTF_2 | HRTIM_DTR_DTF_3;		
// Deadtime falling = 15*Ttg = 104 ns
	HRTIM1->sTimerxRegs[0].DTxR  |= HRTIM_DTR_DTFSLK | HRTIM_DTR_DTRSLK;

	HRTIM1->sTimerxRegs[0].SETx1R |= HRTIM_SET1R_PER;															 
 // Event forces the output to active state for TA1
	HRTIM1->sTimerxRegs[0].RSTx1R |= HRTIM_RST1R_CMP1;															
// Event forces the output to inactive state for TA1

	/************************************************
	 *                 HRTIM start
	 ***********************************************/

	HRTIM1->sCommonRegs.OENR |= HRTIM_OENR_TA1OEN | HRTIM_OENR_TA2OEN;							  	
// Enable output PWM for TA1 and TA2

	HRTIM1->sTimerxRegs[0].TIMxCR |= HRTIM_TIMCR_CONT;														     	
// Continuous mode

	HRTIM1->sMasterRegs.MPER = 65000;																		              	
// Period for master timer
	HRTIM1->sMasterRegs.MCR |= HRTIM_MCR_MCEN | HRTIM_MCR_TACEN;												
// Enable counter for Master and timer A

}

void setup() {
InitHRPWM();
HRTIM1->sTimerxRegs[0].CMP1xR = 10;
	
}

void loop() {


}



Here's zip with sketch 
[https://www.dropbox.com/s/ale0m0r0rhuf6bh/STM32F334%20test.zip?dl=0](url)
